src/sinks/file/mod.rs

```diff
@@ -88,98 +98,125 @@ impl PartitionedFileSink {
 
         Some(bytes)
     }
-}
-
-impl Sink for PartitionedFileSink {
-    type SinkItem = Event;
-    type SinkError = ();
 
-    fn start_send(&mut self, event: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
-        if let Some(key) = self.partition_event(&event) {
-            let encoding = self.encoding.clone();
-
-            let partition = self
-                .partitions
-                .entry(key.clone())
-                .or_insert_with(|| File::new(key.clone(), encoding.clone()));
-
-            self.last_accessed.insert(key.clone(), Instant::now());
-
-            partition
-                .start_send(event)
-                .map_err(|error| error!(message = "Error writing to partition.", %error))
-        } else {
-            Ok(AsyncSink::Ready)
-        }
+    fn deadline_at(&self) -> Instant {
+        Instant::now()
+            .checked_add(self.idle_timeout)
+            .expect("unable to compute next deadline")
     }
 
-    fn poll_complete(&mut self) -> Poll<(), Self::SinkError> {
-        let mut all_files_ready = false;
-
-        for (file, partition) in &mut self.partitions {
-            let ready = match partition.poll_complete() {
-                Ok(Async::Ready(())) => true,
-                Ok(Async::NotReady) => false,
-                Err(error) => {
-                    let file = String::from_utf8_lossy(&file[..]);
-                    error!(message = "Unable to flush file.", %file, %error);
-                    true
+    async fn run(&mut self, mut input: Receiver<Event>) -> crate::Result<()> {
+        loop {
+            let what = select(input.next(), self.files.next()).await;
+            match what {
+                Either::Left((None, _)) => {
+                    // If we got `None` - terminate the processing.
+                    debug!("Receiver exausted, terminating the processing loop");
+                    break;
                 }
-            };
-
-            all_files_ready = all_files_ready || ready;
-        }
-
-        for (key, last_accessed) in &self.last_accessed {
-            if last_accessed.elapsed().as_secs() > self.idle_timeout_secs {
-                if let Some(file) = self.partitions.remove(key) {
-                    let file_path = String::from_utf8_lossy(&key[..]);
-                    debug!(message = "Closing file.", file = %file_path);
-                    self.closing.insert(key.clone(), file);
+                Either::Left((Some(event), _)) => {
+                    let path = match self.partition_event(&event) {
+                        Some(path) => path,
+                        None => {
+                            // We weren't able to find the path to use for the
+                            // file.
+                            // This is already logged at `partition_event`, so
+                            // here we just skip the event.
+                            warn!("Unable to partition an event, dropping it");
```

Comment:
Most likely want to rate limit this like we do in other sinks.

Reply:
Rate limit the dropping? Could you link some relevant code or describe the desired logic here?

Reply:
Oh we can rate limit how often a log is logged https://github.com/timberio/vector/search?q=rate_limit_secs&unscoped_q=rate_limit_secs